Casual Clustering id reuse

klaren opened the pull request.

klaren committed:
Extracted file handling of reusable and max id into separate class
Updated IdGenerator to use this
Adapted current tests

klaren committed:
Moved highId responsibility to IdGenerator

klaren committed:
Renaming and cleanup

klaren committed:
Moved tests related to filesystem to IdFile instead of IdGeneratorImpl

klaren committed:
Introduce IdReusabilityCondition to keep track of what ids to reuse.
It's only safe to reuse when:
1. You are leader
2. The id is freed in a transaction you created

klaren committed:
Renamed IdFile to IdContainer

klaren committed:
Removed unnecessary test

klaren commented:
Can you guys please take a look at this @digitalstain @martinfurmanski ?

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/id/IdReusabilityCondition.java:
```diff
@@ -0,0 +1,113 @@
+            this.transactionIdStore = transactionIdStore;
+
+            // Get highest transaction id seen
+            this.idHighestWhenBecameLeader = transactionIdStore.getLastCommittedTransactionId();
```
At this point there could be another member thinking it is the leader as well which is happily freeing IDs > transactionIdStore.getLastCommittedTransactionId() of this local instance. Leadership is eventually consistent state whilst lastCommittedTransactionId is a projection of the asynchronous transaction application process (ordered after but asynchronous to the commit path). The best and easiest way to get correct behaviour here is to order the transaction and resulting free IDs in the consensus log relative something else also in the consensus log.

klaren replied:
Since the transactions are in the consensus log and the new leader uses them to prove it has the longest log, the highest seen transactions IDs should correlate with a leader switch. If there is any contention about the leadership the transactions belonging to older terms will not be accepted since only leaders are allowed to append new transactions. 

I don't see how

> thinking it is the leader

is enough to get new transactions committed in the consensus log or how you will accept new transaction from another leader during this uncertain state.

chrisvest replied:
@martinfurmanski Are you still seeing a problem here, or is it good?

MishaDemianenko replied:
we will have a talk with guys next week when Chris will return from vacation, maybe we will need to get highest transaction id from raft log instead

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/id/ReplicatedIdGenerator.java:
```diff
@@ -34,27 +39,38 @@
+        if ( freeIdCondition.getAsBoolean() )
+        {
+            idContainer.freeId( id );
+        }
```
Do we have a feel for the probability of ending up here, and the condition is false?

MishaDemianenko replied:
not sure, I fully understand your question. But in case I am: for period when new master catching up with transaction which already in the raft log but not applied yet

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/id/IdGeneratorImpl.java:
```diff
@@ -64,31 +60,15 @@
-    private final boolean aggressiveReuse;
-    private FreeIdKeeper keeper;
+    private final IdContainer idContainer;
+    private long highId = INTEGER_MINUS_ONE;
```
No need to initialise it to this strange constant because the constructor will overwrite it.

chrisvest reviewed (commented):
Had one more question.

klaren committed:
Switched from using transaction ids to work on raft command indexes,
since transaction id are async and can potentially be behind of actual
member progress

MishaDemianenko commented:
@martinfurmanski @digitalstain guys, can you please take a look into updated version that based on command ids instead, thx!

martinfurmanski commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/scenarios/ClusterIdReuseIT.java:
```diff
@@ -0,0 +1,178 @@
+
+    private void removeTwoNodes( Cluster cluster, MutableLong first, MutableLong second ) throws Exception
+    {
+        cluster.coreTx( ( db, tx ) ->
```
You can extract the member on which this transaction was performed for use later on in the test. This is useful to protect against flaky behaviour. If you try to rely on leadership being stable, then you can and will experience flakiness. This comment applies to any place where you assume that the roles will not change.

An ill-timed leader-switch could also make it so that IDs are actually not reused by anyone, so asserting on that they are will be flaky by itself then. Imagine for example the leader stepping down due to GC right after the transaction got committed in consensus-land. It will eventually be applied but there will be no leader or a new (or the same) leader with a higher lastLogIndexBeforeWeBecameLeader.

A robust assertion that you could make is that IDs are reused on at most one member, but perhaps that is not interesting enough.

A way to make it more robust would be to run some sequence many times, e.g. many delete transactions, and assert that at least some percentage was in fact reused and on the expected members.

martinfurmanski reviewed (commented):
I have mostly looked at the causal clustering parts.

klaren committed:
Removed unnecessary assignment.
Check leaders during test to discard unpredictable leader switches.

klaren committed:
Just for tests

MishaDemianenko committed:
Initialise id controller for ha edition

MishaDemianenko committed:
Cleanup of unused fields, visibility modifiers.
Change what factory is wrapped in core edition module.

MishaDemianenko committed:
Refactor way how we construct id components over editions.

Solve the problem with non clear component initialization sequences and
lost buffered id generator factory got lost.

MishaDemianenko committed:
Initialize id controller in replica edition.

MishaDemianenko committed:
Add additional tests for id components constructed during edition creation.
Update dependency resolver to include only one id generation factory.

MishaDemianenko commented:
@chrisvest @martinfurmanski @digitalstain guys you are welcome to take another round of review!

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/EditionModule.java:
```diff
@@ -68,6 +74,9 @@
 public abstract class EditionModule
 {
+    protected static final boolean safeIdBuffering = FeatureToggles.flag(
+            EditionModule.class, "safeIdBuffering", true );
```
Do we still need the feature toggle?

chrisvest replied:
Hmm… maybe still good to have in case there's a field fire.

chrisvest replied:
In that case maybe add a comment that this toggle was placed in RecordStorageEngine in versions prior to 3.3.

chrisvest commented:
Not related to these changes, but one thing that's curious is that relationship group records are not aggressively reused in Enterprise Edition, according to `EnterpriseEditionSettings.idTypesToReuse`, and `CommunityIdTypeConfigurationProvider`.

MishaDemianenko commented:
@digitalstain @martinfurmanski do you guys have anything to add or we good to go?

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/id/IdReusabilityCondition.java:
```diff
@@ -0,0 +1,103 @@
+                oldTransactionsApplied = true;
+            }
+
+            return hasAppliedOldTransactions;
```
This code can be simplified. Would be nice with a comment describing the deeper meaning behind this logic.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/id/IdContainer.java:
```diff
@@ -0,0 +1,283 @@
+import org.neo4j.kernel.impl.store.UnderlyingStorageException;
+
+
+public class IdContainer
```
Would be nice with some javadocs on this. For instance, it doesn't look thread-safe, so that might be a good thing to note.

chrisvest commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/id/ReplicatedIdGenerator.java:
```diff
@@ -35,26 +39,31 @@
     @Override
     public void freeId( long id )
     {
+        idContainer.freeId( id );
```
`IdGeneratorImpl` protects its `IdContainer` by making its methods `synchronized`. This class does not do that, so does this not expose the `IdContainer` to concurrency that it is not prepared to handle?

klaren replied:
Nice catch! BufferedIdGenerator runs in a separate thread when freeing IDs so this is definitely a problem.

chrisvest reviewed (commented):
I'm not that familiar with the causal clustering stuff. I had a few comments.

klaren committed:
Cleanup and simplification after review

klaren commented:
@chrisvest Done with changes. Thanks for the feedback!

MishaDemianenko committed:
Checkstyle fix after rebase
